Multiple root components start in parallel.

MackinnonBuck opened the pull request:
**Multiple root components start in parallel.**
Each root component is added and processed in parallel, rather than waiting for one to become quiescent before adding the next.

**PR Description**
Hosts now add each root component ignoring the task awaiting quiescence. Then, after all root components are added, the host awaits quiescence for all components.

Addresses #34306

MackinnonBuck committed:
Initial prototype.

MackinnonBuck commented on src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyHost.cs:
```diff
@@ -161,11 +161,16 @@ internal async Task RunAsyncCore(CancellationToken cancellationToken, WebAssembl
                         foreach (var rootComponent in rootComponents)
                         {
-                            await renderer.AddComponentAsync(rootComponent.ComponentType, rootComponent.Parameters, rootComponent.Selector);
+                            _ = renderer.AddComponentAsync(rootComponent.ComponentType, rootComponent.Parameters, rootComponent.Selector);
```
An alternative to doing something like this could be introducing a new method in `Renderer`, maybe `BeginRenderRootComponent`. This would do what `RenderRootComponentAsync` does except await quiescence. Then we could avoid allocating `Task`s we're not even using. This line would then look something like

```suggestion
                            renderer.AddComponent(rootComponent.ComponentType, rootComponent.Parameters, rootComponent.Selector);
```

SteveSandersonMS replied:
I thought RenderRootComponentAsync did already wait for quiescence, but maybe am misremembering (could check on Monday). 

What happens if we collect all the tasks returned here and then Task.WhenAll on them, rather than making WaitForQuiescence public?

MackinnonBuck replied:
@SteveSandersonMS Maybe I didn't word my comment correctly - yes, `RenderRootComponentAsync` does wait for quiescence. I like the `Task.WhenAll` idea, though.

MackinnonBuck replied:
Actually, I'm thinking it still might be better (?) to make an alternative to `RenderRootComponentAsync` that doesn't wait for quiescence, then we wait for quiescence after we've started rendering each component (so, similar to how it currently works but we're not throwing away a bunch of tasks). If we did `Task.WhenAll`, each task being awaited is really the same thing, since `WaitForQuiescence` waits for all pending tasks to complete, and the pending tasks are shared across all components. Is my understanding correct? Do you think `Task.WhenAll` would still be preferable?

SteveSandersonMS replied:
If it were me I'd err on the side of not adding further public API to the renderer, as it doesn't seem necessary given the option of `Task.WhenAll`. I know most of the time all the tasks in the collection will actually be the same instance, but that doesn't seem like a drawback to me - it's just how it needs to be to cover all possible scenarios.

MackinnonBuck commented on src/Components/Server/src/Circuits/CircuitHost.cs:
```diff
@@ -105,17 +105,18 @@ public Task InitializeAsync(ProtectedPrerenderComponentApplicationStore store, C
                     await OnCircuitOpenedAsync(cancellationToken);
                     await OnConnectionUpAsync(cancellationToken);
 
-                    // We add the root components *after* the circuit is flagged as open.
-                    // That's because AddComponentAsync waits for quiescence, which can take
-                    // arbitrarily long. In the meantime we might need to be receiving and
-                    // processing incoming JSInterop calls or similar.
```
Now that this isn't the case, would there be any reason to change it so root components are added *before* the circuit is flagged as open?

SteveSandersonMS replied:
I think it still is the case. JS interop calls can happen at any time, including while we're waiting for the first (or all) root component(s) to be added. Or am I misunderstanding your point?

MackinnonBuck replied:
I was more just unsure if this comment was implying that the only reason we're adding root components after the circuit is flagged as open is that `AddComponentAsync` waits for quiescence. So I guess I was wondering, if we're not immediately awaiting the `Task` returned by `AddComponentAsync`, and instead awaiting them all later (via `Task.WhenAll`), would there be any reason then to put the "flag the circuit as open" step in between adding the components and waiting for quiescence? e.g.:

```csharp
for (var i = 0; i < count; i++)
{
  // Call `AddComponentAsync` and collect its task.
}

// Flag the circuit as open.
await OnCircuitOpenedAsync(cancellationToken);
await OnConnectionUpAsync(cancellationToken);

// Await quiescence.
await Task.WhenAll(...);
```

I personally don't see a reason why we would need or want to do this, but I'm just bringing it up because I saw that code comment and was unsure if it was implying something like this would be desirable somehow.

MackinnonBuck replied:
I'm entirely expecting the answer to be "no, we shouldn't do this", but I just want to make sure there's not something I'm glancing over.

SteveSandersonMS replied:
There could be a back-compat argument for keeping the previous ordering, since developers would have been able to observe their OnCircuitOpenedAsync/OnConnectionUpAsync callbacks firing before their component constructors.

If there was a strong reason for changing it we probably could do, but since it seems equally valid either way, I would lean towards retaining the ordering. If you come up with any reason why you think it really should be changed then let us know!

MackinnonBuck committed:
Merge branch 'main' into t-mbuck/parallel-root-components

MackinnonBuck committed:
Update InputDateType.cs

MackinnonBuck committed:
PR feedback

MackinnonBuck committed:
Merge branch 'main' into t-mbuck/parallel-root-components

MackinnonBuck committed:
Removed RootComponentMapping.Deconstruct extension method.

There wasn't a way to make this comply with the DynamicallyAccessedMembers code analysis.

SteveSandersonMS reviewed (approved):
Looks great!

I'm not sure how practical it is to E2E test this. If you think there is a straightforward way then that would be cool, but if it involves adding a load of extra testing infrastructure then I'm not sure the tradeoff would be worth it.

Actually it might be much more reasonable to add unit tests though. I see we already have `CircuitHostTest` - that could probably be extended to cover this new behavior. What do you think?

pranavkm commented on src/Components/Server/src/Circuits/CircuitHost.cs:
```diff
@@ -110,17 +110,19 @@ public Task InitializeAsync(ProtectedPrerenderComponentApplicationStore store, C
+                    // Here, we add each root component but don't await the returned tasks so that the
+                    // components can be processed in parallel.
                     var count = Descriptors.Count;
+                    var pendingRenders = new Task[count];
```
If you'd like this to avoid this allocation, you could use an array pool

```C#
var pendingRenders = ArrayPool<Task>.Shared.Rent(count);
try
{
  ...
   await Task.WhenAll(tasks.Take(count));
}
finally
{
   ArrayPool<Task>.Shared.Return(tasks);
}

SteveSandersonMS replied:
Is that really worth it though? This only happens once per circuit startup, and circuit startup involves (probably) hundreds of allocations. Do we really want to have to reason about the correctness of pooling logic?

pranavkm replied:
> Is that really worth it though?

Kinda why I didn't insist on it. It's a nice to have, but I wouldn't think this is critical especially in the absence of profiling data.

MackinnonBuck committed:
Now using ArrayPool to avoid allocations.

MackinnonBuck committed:
Revert "Now using ArrayPool to avoid allocations."

This reverts commit 420f73b8ce82d425f45740c709fb1ecd1c54b690.

MackinnonBuck committed:
Added test for CircuitHost rendering root components in parallel.

MackinnonBuck committed:
Update CircuitHostTest.cs

MackinnonBuck committed:
PR feedback.
